feature/waku-rpc (#135)
vacp2p/rfc: specs/waku/waku-rpc-api.md

---
title: Waku RPC API
version: 1.0.0

> status: Draft

Reviewer: Is this a Draft? Vs stable

If draft version seems off

Also should other specs be updated to point to this?

Reviewer: ^

Reviewer: re other specs

Author: Where should we point to this? In the waku spec?

authors: Dean Eigenmann <dean@status.im>
---

## Table of Contents

1. [Abstract](#abstract)
2. [Wire Protocol](#wire-protocol)
    1. [Transport](#transport)
    1. [Objects](#objects)
    1. [Methods](#methods)
3. [Copyright](#copyright)

## Abstract

> In this specification we describe the RPC API that Waku nodes SHOULD adhere to. The unified API allows clients to easily

Reviewer: Might be worth mentioning somewhere that this is a privileged API and not like some REST API, since you store keys for filter

Reviewer 2: Yes, I would also change SHOULD to MAY. As it is really something for a client to decide on imo.
We could even add some sentence that this typically would be defaulted to access from localhost only.

> In this specification we describe the RPC API that Waku nodes MAY adhere to. The unified API allows clients to easily
> be able to connect to any node implementation. The API described is privileged as a node stores the keys of clients. 

Reviewer 3: ```suggestion
This specification describes the RPC API that Waku nodes MAY adhere to. The unified API allows clients to easily
be able to connect to any node implementation. The API described is privileged as a node stores the keys of clients. 
```

## Introduction

This API is based off the [Whisper V6 RPC API](https://github.com/ethereum/go-ethereum/wiki/Whisper-v6-RPC-API).

## Wire Protocol

### Transport

> Nodes SHOULD expose a [JSON RPC](https://www.jsonrpc.org/specification) API that can be publicly accessed. The JSON RPC version SHOULD be `2.0`. Below is an example request:

Reviewer: Publicly accessible, what do you mean? I'm not sure that's accurate.

Reviewer 2: Indeed, see above comment.

```json
{
  "jsonrpc":"2.0",

>   "method":"shh_version",

Reviewer 3: Should this be `waku_version`?

  "params":[],
  "id":1
}
```

#### Fields

| Field     | Description                                         |
| --------- | --------------------------------------------------- |
| `jsonrpc` | Contains the used JSON RPC version (`Default: 2.0`) |
| `method`  | Contains the JSON RPC method that is being called   |
| `params`  | An array of parameters for the request              |
| `id`      | The request ID                                      |

### Objects

In this section you will find objects used throughout the JSON RPC API.

#### Message

> The message object represents a Waku message. Below you will find the description of the attributes contained in the message object.

Reviewer 2: Perhaps it should be stated what a Waku message is (considering the previous issue with message vs envelope usage).
We could define here the term `Waku message` e.g. as the decrypted payload (and padding) of the Waku envelope + its useful metadata and extra info (such as hash).

| `topic` | string | 4 bytes, the message topic |
| `payload` | string | Decrypted payload |
| `padding` | string | Optional padding, byte array of arbitrary length |
| `pow` | number | The proof of work value |
| `hash` | string | Hash of the enveloped message |

#### Filter

The filter object represents filters that can be applied to retrieve messages. Below you will find the description of the attributes contained in the filter object.

| Field | Type | Description |
| ----: | :--: | ----------- |
| `symKeyID` | string | ID of the symmetric key for message decryption |
| `privateKeyID` | string | ID of private (asymmetric) key for message decryption |
| `sig` | string | Public key of the signature |
| `minPow` | number | Minimal PoW requirement for incoming messages |
| `topics` | array | Array of possible topics, this can also contain partial topics |
| `allowP2P` | boolean | Indicates if this filter allows processing of direct peer-to-peer messages |

All fields are optional, however `symKeyID` or `privateKeyID` must be present, it cannot be both. Additionally, the `topics` field is only optional when an asymmetric key is used.

> ### Methods

Reviewer: What accounts for the difference in Nim and Go impl? Is this the minimal subset? Are there yet undocumented ones that we want here?

Author: So this is based off of whisper v6 which matches what is in Waku Nim, I am not sure on the go implementation.

Reviewer 3: FYI https://github.com/status-im/status-go/blob/develop/waku/api.go has all the Waku RPC methods currently supported by `status-go`

I notice that `Version()` is missing

#### `waku_version`

The `waku_version` method returns the current version number.

##### Parameters

none

##### Response

- **string** - The version number.

#### `waku_info`

The `waku_info` method returns information about a Waku node.

##### Parameters

none

##### Response

The response is an `Object` containing the following fields:

- **`minPow` [number]** - The current PoW requirement.

> - **`maxMessageSize` [float]** - The current maximum message size in bytes.

Reviewer 2: Considering comment below, perhaps we should consider renaming this also as size check is done on envelope size.

- **`memory` [number]** - The memory size of the floating messages in bytes.

> - **`messages` [number]** - The number of floating messages.

Reviewer 2: If we would like to be really correct, I guess that this should be `envelopes`

> #### `waku_setMaxMessageSize`

Reviewer 2: and this.

> Sets the maximum envelope size allowed by this node. Any envelopes larger than this size both incoming and outgoing will be rejected. The envelope size can never exceed the underlying envelope size of `10mb` <!-- is this still accurate -->

Reviewer 4: Is this still inaccurate?

##### Parameters

 - **number** - The message size in bytes.

##### Response

- **bool** - `true` on success or an error on failure.

#### `waku_setMinPoW`

Sets the minimal PoW required by this node.

##### Parameters

 - **number** - The new PoW requirement.

##### Response

- **bool** - `true` on success or an error on failure.

#### `waku_markTrustedPeer`

Marks a specific peer as trusted allowing it to send expired messages.

##### Parameters

- **string** - `enode` of the peer.

##### Response

- **bool** - `true` on success or an error on failure.

#### `waku_newKeyPair`

> Generates a public and private key pair used for message encryption and decryption.

Reviewer: public private key or keypair

##### Parameters

none

##### Response

 - **string** - The key ID or an error on failure.

#### `waku_addSymKey`

Stores the key and returns its ID.

##### Parameters

- **string** - The raw key for symmetric encryption hex encoded.

##### Response

 - **string** - The key ID or an error on failure.

#### `waku_generateSymKeyFromPassword`

Generates the key from a password and stores it.

##### Parameters

 - **string** - The password.

##### Response

 - **string** - The key ID or an error on failure.

#### `waku_hasSymKey`

Returns whether there is a key associated with the ID.

##### Parameters

- **string** - ID of the Key.

##### Response

- **bool** - `true` or `false` or an error on failure.

#### `waku_getSymKey`

Returns the symmetric key associated with an ID.

##### Parameters

- **string** - ID of the Key.

##### Response

- **string** - Raw key on success or a failure on error.

#### `waku_deleteSymKey`

Deletes the key associated with an ID.

##### Parameters

- **string** - ID of the Key.

##### Response

- **bool** - `true` or `false` or an error on failure.

#### `waku_subscribe`

> Creates and registers a new subscription to receive notifications for inbound waku messages.

Reviewer 2: ```suggestion
Creates and registers a new subscription to receive notifications for inbound Waku messages.
```

##### Parameters

The parameters for this request is an array containing the following fields:

>  1. **string** - The ID of the function call, in case of waku this must contain the value "messages".

Reviewer 2: ```suggestion
 1. **string** - The ID of the function call, in case of Waku this must contain the value "messages".
```

>  2. **object** - The [message filter](#message-filter).

Reviewer 2: ```suggestion
 2. **object** - The [message filter](#filter).
```

###### Notifications

Notifications received by the client contain a [message](#message) matching the filter. Below is an example notification:

```json
{
  "jsonrpc": "2.0",
  "method": "shh_subscription",
  "params": {
    "subscription": "02c1f5c953804acee3b68eda6c0afe3f1b4e0bec73c7445e10d45da333616412",
    "result": {
      "sig": "0x0498ac1951b9078a0549c93c3f6088ec7c790032b17580dc3c0c9e900899a48d89eaa27471e3071d2de6a1f48716ecad8b88ee022f4321a7c29b6ffcbee65624ff",
      "recipientPublicKey": null,
      "ttl": 10,
      "timestamp": 1498577270,
      "topic": "0xffaadd11",
      "payload": "0xffffffdddddd1122",
      "padding": "0x35d017b66b124dd4c67623ed0e3f23ba68e3428aa500f77aceb0dbf4b63f69ccfc7ae11e39671d7c94f1ed170193aa3e327158dffdd7abb888b3a3cc48f718773dc0a9dcf1a3680d00fe17ecd4e8d5db31eb9a3c8e6e329d181ecb6ab29eb7a2d9889b49201d9923e6fd99f03807b730780a58924870f541a8a97c87533b1362646e5f573dc48382ef1e70fa19788613c9d2334df3b613f6e024cd7aadc67f681fda6b7a84efdea40cb907371cd3735f9326d02854",
      "pow": 0.6714754098360656,
      "hash": "0x17f8066f0540210cf67ef400a8a55bcb32a494a47f91a0d26611c5c1d66f8c57"
    }
  }
}
```

#### `waku_unsubscribe`

> Cancels and removes an existing subscription.

Reviewer: What does this correspond to in actions the node should take?

##### Parameters

 - **string** - The subscription ID.

##### Response

- **bool** - `true` or `false`

#### `waku_newMessageFilter`

Creates a new message filter within the node. This filter can be used to poll for new messages that match the criteria.

##### Parameters

> The request must contain a [message filter](#message-filter) as its parameter. 

Reviewer 2: ```suggestion
The request must contain a [message filter](#filter) as its parameter. 
```

##### Response

 - **string** - The ID of the filter.

#### `waku_deleteMessageFilter`

Removes a message filter from the node.

##### Parameters

- **string** - ID of the filter created with [`waku_newMessageFilter`](#waku_newMessageFilter).

##### Response

- **bool** - `true` on success or an error on failure.

#### `waku_getFilterMessages`

Retrieves messages that match a filter criteria and were received after the last time this function was called.

##### Parameters

- **string** - ID of the filter created with [`waku_newMessageFilter`](#waku_newMessageFilter).

##### Response

The response contains an array of [messages](#messages) or an error on failure.

#### `waku_post`

> The `waku_post` method creates a [waku message](./waku-1.md#messages) and disseminates it to the network.

Reviewer 2: Here we should call it envelope. And perhaps link to envelope or say it is send through the messages packet over the network and keep the current link.

##### Parameters

The parameters is an `Object` containing the following fields:
 - **`symKeyID` [string]** `optional` - The ID of the symmetric key used for encryption
 - **`pubKey` [string]** `optional` - The public key for message encryption.
 - **`sig` [string]** `optional` - The ID of the signing key.
 - **`ttl` [number]** - The time-to-live in seconds.
 - **`topic` [string]** - 4 bytes message topic.
 - **`payload` [string]** - The payload to be encrypted.
 - **`padding` [string]** `optional` - The padding, a byte array of arbitrary length.
 - **`powTime` [number]** - Maximum time in seconds to be spent on the proof of work.
 - **`powTarget` [number]** - Minimal PoW target required for this message.
 - **`targetPeer` [string]** `optional` - The optional peer ID for peer-to-peer messages.
 
*Either the **`symKeyID`** or the **`pubKey`** need to be present. It can not be both.*

#### Response

> - **bool** - `true` on success or an error on failure.

Reviewer 2: I think the status-go implementation returns the hash of the envelope here. At least it used to for whisper.
nim-waku does just return true, but I guess the hash could be useful.

Author: I am fine with keeping it or changing it.

> The `waku_post` method creates a waku envelope and disseminates it to the network.

Reviewer 4: propagates?
